### PR TITLE
many: allow 'system-usernames' with libseccomp > 2.4 and golang-seccomp > 0.9.0

### PR DESCRIPTION
**Pull request by jdstrand**
This is broken out from https://github.com/snapcore/snapd/pull/6681 and builds upon https://github.com/snapcore/snapd/pull/7111 and implements the libseccomp and golang-seccomp checks starting with 5ce5ff1.

### CODE REVIEW

**Comment by jdstrand** [comment]
Marking as Blocked for now since it is on top of 7111 and it is not ready for review yet.

**Comment by jdstrand** [comment]
7111 is now unblocked, and therefore so is this.

File: overlord/snapstate/check_snap.go [hunk 1]
```diff
@@ -455,6 +458,51 @@ func checkSystemUsernames(si *snap.Info) error {
...
+	vi, err := interfaces.SeccompCompilerVersionInfo(filepath.Join(filepath.Dir(path), "snap-seccomp"))
+	if err != nil {
+		return fmt.Errorf("Could not obtain seccomp compiler information: %v", err)
+	}
+	libseccompVersion, err := seccomp_compiler.VersionInfo(vi).LibseccompVersion()
+	if err != nil {
+		return err
+	}
+
+	// Parse <libseccomp version>
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go]: again feels like LibseccompVersion() should return [3]int or a struct perhaps

File: overlord/snapstate/check_snap.go [hunk 2]
```diff
@@ -455,6 +458,51 @@ func checkSystemUsernames(si *snap.Info) error {
...
+	// fixed, so base the decision on that. ActLog is first available in
+	// 0.9.1.
+	res, err := seccomp_compiler.VersionInfo(vi).HasFeature("bpf-actlog")
+	if err != nil {
+		return err
+	}
+	if !res {
+		return fmt.Errorf(`This snap requires that snapd be compiled against golang-seccomp >= 0.9.1`)
+	}
+
```
Comment (pedronis) [inline overlord/snapstate/check_snap.go]: this overall check should probably be encapsulated on VersionInfo itself as something like SupportsCorrectArgumentFiltering() bool?

File: overlord/snapstate/check_snap.go [hunk 3]
```diff
@@ -455,6 +458,51 @@ func checkSystemUsernames(si *snap.Info) error {
 	if len(si.SystemUsernames) == 0 {
 		return nil
 	}
+
+	// Run /.../snap-seccomp version-info
+	path, err := cmd.InternalToolPath("snapd")
+	if err != nil {
+		return err
+	}
+	vi, err := interfaces.SeccompCompilerVersionInfo(filepath.Join(filepath.Dir(path), "snap-seccomp"))
```
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/check_snap.go]: this is likely follow up material

File: overlord/snapstate/check_snap.go (hunk 1 above)
Comment (pedronis) [reply overlord/snapstate/check_snap.go]: this is also follow up material, might be premature

File: overlord/snapstate/check_snap.go (hunk 2 above)
Comment (pedronis) [reply overlord/snapstate/check_snap.go]: this instead seems the main point, the detail of the check don't belong here, but we want a new dedicated predicate on VersionInfo

File: overlord/snapstate/check_snap.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jdstrand) [reply overlord/snapstate/check_snap.go]: I found at this point in check_snap.go, I don't have anything like InterfaceManager or Backends to call into. I could move all of this into a helper from somewhere else, but went with your main point below which is to call `snap-seccomp version-info` if the snap specifies system-usernames, then call a simple API on the result which removes all the lowlevel details and puts them in sandbox.

**Comment by jdstrand** [comment]
@pedronis - I believe I addressed all feedback.

**Comment by pedronis** [comment]
@jdstrand we merged #7111 after some further fixes, tweaks. I have now merged master into this and adjusted error messages to be consistent with the go/project style

File: overlord/snapstate/check_snap.go [hunk 4]
```diff
@@ -460,6 +463,33 @@ func checkSystemUsernames(si *snap.Info) error {
...
+	}
+
+	vi, err := interfaces.SeccompCompilerVersionInfo(filepath.Join(filepath.Dir(path), "snap-seccomp"))
+	if err != nil {
+		return fmt.Errorf("cannot obtain seccomp compiler information: %v", err)
+	}
+
+	// If the system doesn't support robust argument filtering then we
+	// can't support system-usernames
+	err = seccomp_compiler.VersionInfo(vi).SupportsRobustArgumentFiltering()
```
Comment (mvo5) [inline overlord/snapstate/check_snap.go]: (nitpick^2) This could be a single line: `if err := seccomp_compiler.VersionInfo(vi)...; err != nil {`

**Comment by jdstrand** [comment]
Thanks again for the reviews and polishing! :)

**Pull request merged**